File: libs/backend/resume-python/devfolio_resume_python/postgres_repository.py
"""
libs/backend/resume-python/src/postgres_repository.py

PostgresRepository implements IResumeRepository using asyncpg.
Queries Supabase (or any Postgres) for resume data by variant slug.
"""
from __future__ import annotations

import asyncpg

from .models import (
    Education,
    Profile,
    ProficiencyLevel,
    Project,
    ProjectCategory,
    Skill,
    SkillCategory,
    WorkExperience,
)
from .repository import DEFAULT_RESUME


class PostgresRepository:
    """
    Postgres-backed repository. Receives a live asyncpg connection pool
    and queries the DB for all resume data.
    """

    def __init__(self, pool: asyncpg.Pool) -> None:
        self._pool = pool

    # ── Profile ───────────────────────────────────────────────────────────────

    async def find_profile(self, resume: str = DEFAULT_RESUME) -> Profile:
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                'SELECT * FROM profiles WHERE resume_slug = $1', resume
            )
        if not row:
            raise ValueError(f'No profile found for resume: {resume}')
        return Profile(
            name      = row['name'],
            title     = row['title'],
            clearance = row['clearance'],
            summary   = row['summary'],
            email     = row['email'],
            location  = row['location'],
            github    = row['github'],
            linkedin  = row['linkedin'],
            website   = row['website'],
            avatar_url= row['avatar_url'],
        )

    # ── Work Experience ───────────────────────────────────────────────────────

    async def find_work_experience(self, resume: str = DEFAULT_RESUME) -> list[WorkExperience]:
        async with self._pool.acquire() as conn:
            rows = await conn.fetch('''
                SELECT
                    r.slug      AS role_slug,
                    r.title     AS role_title,
                    r.start_date,
                    r.end_date,
                    r.sort_order AS role_order,
                    c.slug      AS company_slug,
                    c.name      AS company_name,
                    c.location  AS company_location,
                    c.company_url
                FROM roles r
                JOIN companies c ON c.slug = r.company_slug
                ORDER BY r.start_date DESC, r.sort_order ASC
            ''')

            # Fetch all highlights for this resume in one query
            highlight_rows = await conn.fetch('''
                SELECT role_slug, body, sort_order
                FROM highlights
                WHERE resume_slug = $1
                ORDER BY role_slug, sort_order
            ''', resume)

        # Group highlights by role_slug
        highlights: dict[str, list[str]] = {}
        for h in highlight_rows:
            highlights.setdefault(h['role_slug'], []).append(h['body'])

        return [
            WorkExperience(
                id           = row['role_slug'],
                company      = row['company_name'],
                title        = row['role_title'],
                location     = row['company_location'],
                start_date   = row['start_date'],
                end_date     = row['end_date'],
                current      = row['end_date'] is None,
                summary      = '',
                highlights   = highlights.get(row['role_slug'], []),
                technologies = [],
                company_url  = row['company_url'],
            )
            for row in rows
        ]

    # ── Education ─────────────────────────────────────────────────────────────

    async def find_education(self, resume: str = DEFAULT_RESUME) -> list[Education]:
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                'SELECT * FROM education ORDER BY sort_order ASC'
            )
        return [
            Education(
                id          = row['slug'],
                institution = row['institution'],
                degree      = row['degree'],
                field       = row['field'],
                start_date  = row['start_date'],
                end_date    = row['end_date'],
                current     = row['end_date'] is None,
            )
            for row in rows
        ]

    # ── Skills ────────────────────────────────────────────────────────────────

    async def find_skills(self, resume: str = DEFAULT_RESUME) -> list[SkillCategory]:
        async with self._pool.acquire() as conn:
            cat_rows = await conn.fetch('''
                SELECT id, name FROM skill_categories
                WHERE resume_slug = $1
                ORDER BY sort_order ASC
            ''', resume)

            skill_rows = await conn.fetch('''
                SELECT s.*
                FROM skills s
                JOIN skill_categories sc ON sc.id = s.category_id
                WHERE sc.resume_slug = $1
                ORDER BY s.category_id, s.sort_order ASC
            ''', resume)

        # Group skills by category_id
        skills_by_cat: dict[int, list[Skill]] = {}
        for s in skill_rows:
            skill = Skill(
                name                = s['name'],
                proficiency         = ProficiencyLevel(s['proficiency']),
                years_of_experience = s['years_of_experience'],
                highlighted         = s['highlighted'],
            )
            skills_by_cat.setdefault(s['category_id'], []).append(skill)

        return [
            SkillCategory(
                category = cat['name'],
                skills   = skills_by_cat.get(cat['id'], []),
            )
            for cat in cat_rows
        ]

    # ── Projects ──────────────────────────────────────────────────────────────

    async def find_projects(self, resume: str = DEFAULT_RESUME) -> list[Project]:
        async with self._pool.acquire() as conn:
            proj_rows = await conn.fetch(
                'SELECT * FROM projects ORDER BY sort_order ASC'
            )
            tech_rows = await conn.fetch(
                'SELECT * FROM project_technologies ORDER BY project_slug, sort_order ASC'
            )
            hl_rows = await conn.fetch(
                'SELECT * FROM project_highlights ORDER BY project_slug, sort_order ASC'
            )

        techs: dict[str, list[str]] = {}
        for t in tech_rows:
            techs.setdefault(t['project_slug'], []).append(t['name'])

        highlights: dict[str, list[str]] = {}
        for h in hl_rows:
            highlights.setdefault(h['project_slug'], []).append(h['body'])

        return [
            Project(
                id           = row['slug'],
                name         = row['name'],
                description  = row['description'],
                summary      = row['summary'],
                technologies = techs.get(row['slug'], []),
                github_url   = row['github_url'],
                live_url     = row['live_url'],
                featured     = row['featured'],
                start_date   = row['start_date'],
                current      = row['current'],
                highlights   = highlights.get(row['slug'], []),
                category     = ProjectCategory(row['category']),
            )
            for row in proj_rows
        ]

    # ── DB introspection ──────────────────────────────────────────────────────

    async def get_database_version(self) -> str:
        async with self._pool.acquire() as conn:
            return await conn.fetchval('SELECT version()')

    async def get_database_name(self) -> str:
        async with self._pool.acquire() as conn:
            return await conn.fetchval('SELECT current_database()')

File: libs/backend/resume-ruby/lib/devfolio_resume_ruby/postgres_repository.rb
# frozen_string_literal: true

require_relative "repository"

module DevfolioResumeRuby
  class PostgresRepository < Repository
    def get_profile(resume:)
      row = execute_one(<<~SQL, resume)
        SELECT p.name, p.title, p.clearance, p.summary,
               p.email, p.location, p.github, p.linkedin
        FROM profiles p
        WHERE p.resume_slug = $1
      SQL
      return nil unless row

      Profile.new(
        name:      row["name"],
        title:     row["title"],
        clearance: row["clearance"],
        summary:   row["summary"],
        email:     row["email"],
        location:  row["location"],
        github:    row["github"],
        linkedin:  row["linkedin"]
      )
    end

    def get_work_experience(resume:)
      rows = execute_many(<<~SQL, resume)
        SELECT c.name AS company, c.location,
               r.slug AS role_id, r.title, r.start_date, r.end_date,
               h.body AS highlight, h.sort_order AS highlight_order
        FROM companies c
        JOIN roles r ON r.company_slug = c.slug
        LEFT JOIN highlights h ON h.role_slug = r.slug AND h.resume_slug = $1
        ORDER BY r.sort_order, h.sort_order
      SQL

      group_work_experience(rows)
    end

    def get_education(resume:)
      rows = execute_many(<<~SQL)
        SELECT slug, institution, degree, field, start_date, end_date
        FROM education
        ORDER BY sort_order
      SQL

      rows.map do |row|
        Education.new(
          id:          row["slug"],
          institution: row["institution"],
          degree:      row["degree"],
          field:       row["field"],
          start_date:  row["start_date"],
          end_date:    row["end_date"],
          current:     row["end_date"].nil?
        )
      end
    end

    def get_skills(resume:)
      rows = execute_many(<<~SQL, resume)
        SELECT sc.name AS category, sc.sort_order AS cat_order,
               s.name AS skill_name, s.proficiency,
               s.years_of_experience, s.highlighted, s.sort_order
        FROM skill_categories sc
        JOIN skills s ON s.category_id = sc.id
        WHERE sc.resume_slug = $1
        ORDER BY sc.sort_order, s.sort_order
      SQL

      group_skills(rows)
    end

    def get_projects
      rows = execute_many(<<~SQL)
        SELECT p.slug, p.name, p.description, p.summary,
               p.github_url, p.live_url, p.featured,
               p.start_date, p.current, p.category,
               pt.name AS tech,
               ph.body AS highlight, ph.sort_order AS hl_order
        FROM projects p
        LEFT JOIN project_technologies pt ON pt.project_slug = p.slug
        LEFT JOIN project_highlights ph ON ph.project_slug = p.slug
        ORDER BY p.sort_order, pt.sort_order, ph.sort_order
      SQL

      group_projects(rows)
    end

    def get_database_version
      result = ActiveRecord::Base.connection.execute("SELECT version()")
      result.first["version"]
    rescue StandardError => e
      "Error: #{e.message}"
    end

    def get_database_name
      result = ActiveRecord::Base.connection.execute("SELECT current_database()")
      result.first["current_database"]
    rescue StandardError => e
      "Error: #{e.message}"
    end

    private

    def execute_one(sql, *params)
      execute_many(sql, *params).first
    end

    def execute_many(sql, *params)
      if params.any?
        ActiveRecord::Base.connection.exec_query(sql, "SQL", params).to_a
      else
        ActiveRecord::Base.connection.execute(sql).to_a
      end
    end

    def group_work_experience(rows)
      rows.group_by { |r| r["role_id"] }.map do |_role_id, role_rows|
        first = role_rows.first
        highlights = role_rows
          .select { |r| r["highlight"] }
          .sort_by { |r| r["highlight_order"].to_i }
          .map { |r| r["highlight"] }

        WorkExperience.new(
          id:           first["role_id"],
          company:      first["company"],
          title:        first["title"],
          location:     first["location"],
          start_date:   first["start_date"],
          end_date:     first["end_date"],
          current:      first["end_date"].nil?,
          summary:      "",
          highlights:   highlights,
          technologies: []
        )
      end
    end

    def group_skills(rows)
      rows.group_by { |r| r["category"] }.map do |category, skill_rows|
        SkillCategory.new(
          category: category,
          skills:   skill_rows.map do |r|
            Skill.new(
              name:                r["skill_name"],
              proficiency:         r["proficiency"],
              years_of_experience: r["years_of_experience"],
              highlighted:         r["highlighted"]
            )
          end
        )
      end
    end

    def group_projects(rows)
      rows.group_by { |r| r["slug"] }.map do |_slug, proj_rows|
        first = proj_rows.first
        Project.new(
          id:           first["slug"],
          name:         first["name"],
          description:  first["description"],
          summary:      first["summary"],
          technologies: proj_rows.map { |r| r["tech"] }.compact.uniq,
          github_url:   first["github_url"],
          live_url:     first["live_url"],
          featured:     first["featured"],
          start_date:   first["start_date"],
          current:      first["current"],
          highlights:   proj_rows
            .select { |r| r["highlight"] }
            .sort_by { |r| r["hl_order"].to_i }
            .map { |r| r["highlight"] },
          category:     first["category"]
        )
      end
    end
  end
end